236 create thumbnails

raq929 opened the pull request:
## Status

Ready for review

## Description of Changes

Fixes #236 .

Changes proposed in this pull request:

 - Instead of saving image crop data, this PR actually crops the image and uploads a cropped image to s3
 - A temporary migration takes all existing faces and changes them to cropped images
 - the new Faces created keep a FK to the original Image

## Notes for Deployment

**IMPORTANT** 
Immediately after migrations are run, the migration that changes existing faces to images should be removed. (The following migration that removes columns from the face table will need to be updated as well at that time.)

## Screenshots (if appropriate)
![screen shot 2018-06-04 at 4 02 53 pm](https://user-images.githubusercontent.com/13922520/40938790-cb537edc-6810-11e8-9a56-11e59af5cb73.png)


## Tests and linting
 
 [ x] I have rebased my changes on current `develop`
 
 [ x] pytests pass in the development environment on my local machine
 
 [ x] `flake8` checks pass

redshiftzero commented on OpenOversight/app/models.py:
```diff
@@ -102,10 +102,6 @@ class Face(db.Model):
-    face_position_x = db.Column(db.Integer, unique=False)
-    face_position_y = db.Column(db.Integer, unique=False)
-    face_width = db.Column(db.Integer, unique=False)
-    face_height = db.Column(db.Integer, unique=False)
```
hm, so if we ever want to do e.g. #173, it's convenient to have these boxes in the database

redshiftzero commented on OpenOversight/migrations/versions/59e9993c169c_change_faces_to_thumbnails.py:
```diff
@@ -0,0 +1,44 @@
+def upgrade():
+    for face in Face.query.all():
+        # import pdb; pdb.set_trace()
+        if face.face_position_x \
```
so I noticed while testing this migration in the dev env that the test data does not have the bounding box fields filled in. since those columns haven't actually been required (this was a holdover from the early days when we manually made cutouts of pics before submitting them to the website), it's probably worth doing:

```
try:
    if face.face_position_x \ 
    [...]
except AttributeError:
    pass  # then skip this face
```

redshiftzero commented on OpenOversight/tests/test_utils.py:
```diff
@@ -135,3 +136,39 @@ def test_user_cannot_submit_invalid_file_extension(mockdata):
+
+
+# Mock calls to upload_file
+@patch('OpenOversight.app.utils.upload_file', MagicMock(return_value='https://s3-some-bucket/someaddress.jpg'))
```
ah.. that works too :-)

raq929 commented:
OK, this should be good to go. I removed the migration that took away the columns, and added a column to keep track of the original image, since you would also need that data for #173. 
(You'll have to downgrade your db before you pull and then upgrade again afterwards, since I also changed it so that the original image id column is added __before__ the temp migration takes place.)
Also, fyi the creating new faces migration may take a while to run on staging/production, as it involves a fair amount of writing to disk. If we think it will be too much, I can investigate my TODO about using a BytesIO object.

raq929 commented:
@redshiftzero can you be sure to check the migration before pushing to staging? I'm worried about it dropping FKs when it adds the column.

redshiftzero commented:
Noted! and yeah at this stage in the project it's fine if the migration takes a while, I'll just do it on off hours 🌚

redshiftzero commented on OpenOversight/migrations/versions/4a490771dda1_add_original_image_fk.py:
```diff
@@ -0,0 +1,34 @@
+    op.drop_constraint(u'faces_img_id_fkey', 'faces', type_='foreignkey')
+    op.create_foreign_key('fk_face_image_id', 'faces', 'raw_images', ['img_id'], ['id'], onupdate='CASCADE', ondelete='CASCADE', use_alter=True)
+    op.create_foreign_key(None, 'faces', 'raw_images', ['fk_face_original_image_id'], ['id'], onupdate='CASCADE', ondelete='SET NULL', use_alter=True)
+    # ### end Alembic commands ###
```
Verified tags (mapping of face -> officer) are indeed preserved after this migration ✅

redshiftzero commented:
I rebased this one on top of the migration added in #475

Doing a wee bit more investigation - it looks like the reason this migration was hanging for me locally was due to the database being locked:

```
openoversight-dev=> SELECT relation::regclass, * FROM pg_locks WHERE NOT GRANTED;
  relation  | locktype | database | relation | page | tuple | virtualxid | transactionid | classid | objid | objsubid | virtualtransaction |  pid  |      mode       | granted | fastpath 
------------+----------+----------+----------+------+-------+------------+---------------+---------+-------+----------+--------------------+-------+-----------------+---------+----------
 raw_images | relation |    16385 |    16574 |      |       |            |               |         |       |          | 4/14               | 15990 | AccessShareLock | f       | f
 faces      | relation |    16385 |    16654 |      |       |            |               |         |       |          | 3/26               | 15951 | AccessShareLock | f       | f
(2 rows)
```

redshiftzero commented:
Progress, I deployed this to staging by running this migrations separately (as I did _not_ get a db lock when doing this in the dev environment.. 🤷‍♀️ ):

```
python manage.py upgrade db 4a490771dda1
python manage.py upgrade db 59e9993c169c
```

I added one wee change, and then I think this is good to deploy. 

The plan will be to:

1. Backup everything
2. Run each migration separately as above
3. Manually re-tag any images that were added prior to the use of `face_position_x` (#169) - it looks like there are a few images in this situation, but there shouldn't be many
